### PR TITLE
Weighted gap mortar mechanical contact implementation

### LINKED ISSUE #16961: Make mortar contact implementation variationally consistent with literature
## Reason
The use of `NormalNodalLMMechanicalContact` shows improvement in the quality of mechanical contact results. The implementation, as it is, uses a node-wise Lagrange multiplier and its corresponding gap distance to build the contact constraints. As reported in the literature [1], using a node-wise Lagrange multiplier with a gap distance averaged in the mortar sense is a variationally consistent description of mortar mechanical contact constraints. Therefore, adding this combination of nodal LM and mortar gap will ensure MOOSE's formulation consistency. 

## Design
Most likely, there will need to be create somecombination of node-wise constraints (i.e. `NormalNodalLMMechanicalContact`) and mortar averaged classes (i.e. `NormalMortarLMMechanicalContact`). One key aspect of the design will be to ensure that each nodal LM is matched with a gap distance averaged over the right mortar segment.

## Impact
Availability of more widely used mortar contact implementation with variational consistency. In addition, it is possible that the use of averaged gap distance will lead to smoother results when contacting surfaces are not flat -- this is to be determined.

## Reference
[1] Wohlmuth, B., 2011. Variationally consistent discretization schemes and numerical algorithms for contact problems. Acta Numerica, 20, pp.569-734.

### PR DESCRIPTION
**Pull request by lindsayad**
Closes #16961 

@recuero

### CODE REVIEW

File: modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C
Comment (recuero) [inline modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:100]: I am failing to see how the Gauss point variable is changing (maybe it shouldn't) in the calls to the *Qp* methods to fill residuals and Jacobians.

File: modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C
Comment (recuero) [inline modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:144]: Here, the scaling factor applied to the integrated gap corresponds to the primary variables? I.e. `disp_x` or `disp_y`?

File: modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C
Comment (recuero) [inline modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:52]: `_computing_gap_dependence` just refers to computing its Jacobian, doesn't it? When wouldn't we want to use it?

File: modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:53]: If the LM doesn't have any dofs at the node, e.g. a first order LM on a mid-edge node, then we do not apply the constraint because there is no dof to constrain

File: modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:100]: Yea `_qp` is kind of a misnomer here because we are not applying any quadrature rules. We are just evaluating secondary values at one location (the secondary node). (And for `NodeFaceConstraints` which work the the primary face, obviously we only work with one location over there as well...the location where the secondary node projects). But to have consistency with the moose data structures we need to index into our `MooseArrays` holding variable value data; in this case we know the data should have size 1, so `_qp` must be 0. Perhaps it would be more transparent if we didn't do anything with `_qp` and explicitly just indexed with 0?

File: modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:144]: No it's the scaling factor for the LM variable. When we compute the weighted gap in our mortar constraint and save it into the residual vector, the scaling factor for the LM variable gets applied via multiplication. To undo that math, e.g. to make sure that our integrated gap here is truly representative of the gap value and not the gap value times some arbitrary factor, we divide by the scaling factor.

File: modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:150]: Yea this assertion can go. Nice catch!

File: modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:52]: We probably do always want to compute the (approximate) Jacobian. In fact my tests never converged if I didn't have those approximate Jacobians added in.

I copied this logic over from the other class...but yea I think what we want to do is make the `secondary_disp_y` parameter required and then just always compute the gap dependence. :+1:

File: modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C
Comment (recuero) [reply modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:100]: Thanks for the details! I get it now. I think it's fine how it is, just lack of familiarity on my side.

**Review by recuero — DISMISSED** [review]

**Comment by lindsayad** [comment]
@dschwen @bwspenc, could I get a CCB review/approval on this? @recuero approved (before I autosquashed and dismissed his review)

**Comment by recuero** [comment]
@lindsayad Fantastic results! Your weighted gap (`wgap`) mortar implementation generates the best expected outcome!

The image below is for a non-flat geometry, which happens to be circular. I can see two differentiating aspects over node on face and the nodal gap mortar enforcement:

- In an area where there is a geometric defect, the weighted gap reduces the maximum pressure, which makes sense because the finer we discretize the more nodes will share the contacting force. I think this is a feature of mortar that guarantees the passage of the contact patch test. There is a spatial shift, but I think this is probably fine and due to segment-matching effects.

- In a "regular" area where node on face and the old nodal LM mortar formulations generate an artificial and ugly kink in the  results, the weighted gap generates a flawless evolution of the contact force.

The distribution of the contact pressure around the defect may be improved if we use the mixed approach, I'll keep trying combinations.

Thanks for working on this!


image

**Comment by lindsayad** [comment]
Very cool! What is the difference between "node-segment" and "node-segment kinematic"? It looks like node-segment (kinematic) is less oscillatory at the geometric defect than the mortar implementations. I would indeed be curious whether the mixed approach helps with that. It's known that saddle point problems (which you would have in a same-order LM approach when you are in contact) can be oscillatory/produce checkerboards. @jwpeterson showed a lot of this same-order instability in his technical report when looking at solution continuity, which is also saddle-point.

**Comment by recuero** [comment]
"node-segment" is obtained using a penalty approach, as opposed to the kinematic approach. It doesn't add much info to the plot since the results are virtually identical.

**Comment by recuero** [comment]
Using first-order LMs on a second-order mesh (`mixed`) does show improvement with the oscillatory behavior around the defect. But it makes convergence a lot more challenging, to the point that the simulation can't finish. May be related to the approximate AD Jacobians for mechanical contact? This behavior was already present in the previous mortar approaches.

image

**Comment by lindsayad** [comment]
I could believe it's a Jacobian problem. Is it taking a lot of linear iterations? Would you share the inputs you are using for this (either here or privately) or paste the output of a failed solve with `-ksp_monitor_true_residual`?

**Comment by lindsayad** [comment]
@recuero please try playing around with these most recent commits. Your input that you sent me converged within 3 nonlinear iterations for every timestep with SMP.

**Comment by lindsayad** [comment]
You'll probably find some new issues, but I'm pretty stoked with how that input solved! I always like to think that we're about to usher in a new era of mortar...

**Comment by lindsayad** [comment]
After 484dbef the mixed basis mortar test now has the new record for fewest nonlinear iterations to solve the bouncing block problem at 93.

**Comment by lindsayad** [comment]
Mmm I guess that it was actually 93 before that commit too

**Comment by recuero** [comment]
For a relatively fine problem, with the two commits you added yesterday (including the tighter tolerance), I see that:

- For a first order LM on a first order mesh, the number of nonlinear iterations go from 142 to 137. The number of linear iterations go from 1113 to 1028. All this points to improvement on first order meshes.
- For a first order LM on a second order mesh, I couldn't solve the problem without those two commits. With those two commits, the number of nonlinear iterations is 121, and the number of linear iterations is 1450, which I think it's pretty good for this problem.

**Comment by recuero** [comment]
Now, even for a second-order mesh, the thermomechanical problem can be solved to tight tolerances. Big improvement!

**Comment by lindsayad** [comment]
Great news!

**Comment by tophmatthews** [comment]
Are you using thermal/mechanical mortar @recuero ?

**Comment by recuero** [comment]
Yes. For this particular problem, the thermal part wasn't very problematic.

**Comment by lindsayad** [comment]
> For a relatively fine problem, with the two commits you added yesterday (including the tighter tolerance), I see that:
> 
>     * For a first order LM on a first order mesh, the number of nonlinear iterations go from 142 to 137. The number of linear iterations go from 1113 to 1028. All this points to improvement on first order meshes.
> 
>     * For a first order LM on a second order mesh, I couldn't solve the problem without those two commits. With those two commits, the number of nonlinear iterations is 121, and the number of linear iterations is 1450, which I think it's pretty good for this problem.

Pretty interesting, but not totally unexpected to me, that the mixed-basis (second order disp, first order LM) shows less nl its (121) than the equal-order first-order scheme (137). This is consistent with my experience with other saddle-point problems in which the LM has to be of a different order than the primal var to have stability (unless you implement another method of stabilization, e.g. adding additional terms like is done with PSPG for incompressible Navier-Stokes and like we implement for mortar solution continuity) in the solution...and generally what I see is that stability in the solution (e.g. lack of or less oscillations) equates to stability in the (non)linear solve.

**Comment by lindsayad** [comment]
@dschwen @bwspenc I'm pretty happy with where we've gotten to, so feel free to review when able

File: modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C
Comment (fdkong) [inline modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:49]: If we do such an assumption, do we have a error check for users' input?

File: modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i
Comment (fdkong) [inline modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i:59]: it is a bit confusing we do not call `secondary_disp_x`

File: modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i
Comment (fdkong) [inline modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i:66]: Should we make these names consistent?

```

    primary_boundary = 20
     secondary_boundary = 10
```

File: modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i
Comment (fdkong) [inline modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i:82]: I am trying to understand why we still need `normal_x` and `normal_y`.  I feel `weighted_gap_lm` and `ncp_lm` are enough

**Review by fdkong — CHANGES_REQUESTED** [review]
It looks good to me. I am mainly confused by the constraint block of input files.

When applying the weighted gap formula, why we still need `NormalMortarMechanicalContact`?

File: framework/src/constraints/AutomaticMortarGeneration.C
Comment (lindsayad) [reply framework/src/constraints/AutomaticMortarGeneration.C:784]: Haha it is a libMesh macro variable, but I will make this explicit

Comment (lindsayad) [reply framework/src/constraints/AutomaticMortarGeneration.C:784]: I use non-squared `TOLERANCE` in other places

File: modules/contact/doc/content/source/constraints/ApplyPenetrationConstraintLMMechanicalContact.md
Comment (lindsayad) [reply modules/contact/doc/content/source/constraints/ApplyPenetrationConstraintLMMechanicalContact.md:15]: No, slave -> secondary, and master -> primary.

File: modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:28]: Good call. We actually do not need these parameters anymore

File: modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:148]: > Did you mean the current residual vector already stored "the weighted gap"?

Correct!

File: modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:49]: Good idea. I will add error checks

File: modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i
Comment (lindsayad) [reply modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i:66]: So `ApplyPenetrationConstraintLMMechanicalContact` is actually a `NodeFaceConstraint` and it has has the `secondary` and `primary` params for many years. It is fine for them to do this because they don't also have lower-dimensional subdomains associated with their object like `MortarConstraints` do. So while the discrepancy is obnoxious I don't think we should change it at this point. I will note that our users of mortar contact (and contact in general) generally go through the `ContactAction` so this discrepancy is hidden from them

File: modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i
Comment (lindsayad) [reply modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i:59]: I have made changes so that the param name is now `disp_x`

File: modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i
Comment (lindsayad) [reply modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i:82]: The `ComputeWeightedGapLMMechanicalContact` and `ApplyPenetrationConstraintLMMechanicalContact` objects together determine the value of the LM, e.g. they compute the LM residual. However, they do not apply the LM (which represents the contact pressure) to the displacement residuals/equations. That is what `normal_x` and `normal_y` do. They apply the contact force/pressure represented by the LM to the displacement equations.

**Comment by lindsayad** [comment]
This'll teach me to add more sanity checks!!!

**Comment by lindsayad** [comment]
Ah, beautiful pass again. @fdkong this is ready for further review

**Review by fdkong — APPROVED** [review]

**Pull request merged**
